migrations/2023-06-07-105918_add_hot_rank_columns/up.sql

```diff
@@ -0,0 +1,51 @@
+-- This converts the old hot_rank functions, to columns
+
+-- Remove the old compound indexes
+DROP INDEX idx_post_aggregates_featured_local_newest_comment_time;
+DROP INDEX idx_post_aggregates_featured_community_newest_comment_time;
+DROP INDEX idx_post_aggregates_featured_local_comments;
+DROP INDEX idx_post_aggregates_featured_community_comments;
+DROP INDEX idx_post_aggregates_featured_local_hot;
+DROP INDEX idx_post_aggregates_featured_community_hot;
+DROP INDEX idx_post_aggregates_featured_local_score;
+DROP INDEX idx_post_aggregates_featured_community_score;
+DROP INDEX idx_post_aggregates_featured_local_published;
+DROP INDEX idx_post_aggregates_featured_community_published;
+DROP INDEX idx_post_aggregates_featured_local_active;
+DROP INDEX idx_post_aggregates_featured_community_active;
+
+DROP INDEX idx_comment_aggregates_hot;
+
+DROP INDEX idx_community_aggregates_hot;
+
+-- Add the new hot rank columns for post and comment aggregates
+-- Note: 1728 is the result of the hot_rank function, with a score of 1, posted now
+-- hot_rank = 10000*log10(1 + 3)/Power(2, 1.8)
+alter table post_aggregates add column hot_rank integer not null default 1728;
+alter table post_aggregates add column hot_rank_active integer not null default 1728;
+
+alter table comment_aggregates add column hot_rank integer not null default 1728;
+
+alter table community_aggregates add column hot_rank integer not null default 1728;
+
+-- Populate them initially
+-- Note: After initial population, these are updated in a periodic scheduled job, 
+-- with only the last week being updated.
+update post_aggregates set hot_rank_active = hot_rank(score::numeric, newest_comment_time_necro);
+update post_aggregates set hot_rank = hot_rank(score::numeric, published);
+update comment_aggregates set hot_rank = hot_rank(score::numeric, published);
+update community_aggregates set hot_rank = hot_rank(subscribers::numeric, published);
+
+-- Create single column indexes
+create index idx_post_aggregates_score on post_aggregates (score desc);
+create index idx_post_aggregates_published on post_aggregates (published desc);
+create index idx_post_aggregates_newest_comment_time on post_aggregates (newest_comment_time desc);
+create index idx_post_aggregates_newest_comment_time_necro on post_aggregates (newest_comment_time_necro desc);
+create index idx_post_aggregates_featured_community on post_aggregates (featured_community desc);
+create index idx_post_aggregates_featured_local on post_aggregates (featured_local desc);
```

Comment:
My one concern here: the indexes I had before were compound indexes that had these `featured` columns in the indexes first, since they're always at the top. I'm not sure how the performance will be now that these are all single column.

Reply:
cc @johanndt

Reply:
Honestly the only way to know is to actually run it and compare. But I think performance will be just fine. If you have queries which filter on `newest_comment_time` or `newest_comment_time_necro` WITHOUT the `featured` columns then they can now use the indexes too and will be faster.

Reply:
I'm testing these out now with production data.

[Here's one](http://sprunge.us/hCak6X) of the post table, sorted by `featured_local desc, hot_rank desc`, without a compound index.

Reply:
@Nutomic I've also just added a script to test this:

```
cd scripts/query_testing
./post_query_hot_rank.sh
```

Reply:
I had a look again @dessalines. It's definitely better - the sorting is not the limiting factor anymore, even though it seems that way to you. (when you take the ordering out, it's faster).

Looking at the query plan, it's all the different joins adding up to a slow time. It's returning 35,507 rows. And it's a lot of work to join all those to different tables.

But why is it so fast when you take out the `order by` clauses? Well, you have `LIMIT 40` in there, and if you don't care about the order, it can grab the first 40 posts it finds (in any order), and only do the other joins for 40 posts.

But as soon as you put in a sort condition (as you found out any sort order - not just hot_rank), it means it needs to get all 35,507 rows with the joined values and can only sort right at the end. Because it can only applies the `limit 40` AFTER sorting.

So my suggestion would be to try and limit the number of rows as soon as possible, and join to other rows as late as possible.

For example in shortened pseudo code if you can do this instead:

```
SELECT ...
...
...
FROM (
SELECT * from post INNER JOIN post_aggregate ON .....
WHERE post.featured_local = true AND .....
ORDER BY post_aggregate.hot_rank LIMIT 40
)
LEFT JOIN ...
LEFT JOIN all the other columns
```

If you can rewrite it that post and post_aggregate are joined and sorted first, then you can grab the top 40 there, and only do all these other joins afterwards. Which should only join 40 rows instead of 350,000.

It might not quite be as simple. You might have to join to the ban lists and things as well first to make sure you get the correct 40. But the idea is to join only to what you need to filter, then you limit, and then you join to the rest of the info.

If you need to join to too many other things first, then we'll have to think of some other ideas, like batch updating all the filter conditions onto post_aggregate (like we did with hot_rank).

Reply:
Whoa thanks, this is hugely helpful again! So SQL has been doing lots of work joining thousands of rows *before* the filter / limit, because of the way I've constructed the query.

I'll add an issue for this one specifically, and copy this text. It *should* be doable with diesel, our ORM. 

Thanks!